EPMLSTRCMW-142: Get project versions form gitlab

AndreiMaksimenko opened the pull request:
Add method to get project versions from GilLab
refactor Project Version class and add Commit class
add Json serialization

AndreiMaksimenko committed:
feat: get project versions

Add method to get project versions from GilLab
refactor Project Version class and add Commit class
add Json serialization

Dmitrii-Shevchenko commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -30,13 +30,34 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
+            if (resp.status != StatusCodes.OK.intValue)
+              Left(VersioningException(s"Could not take versions. Response status: ${resp.status}"))
+            else {
+              println(resp.body)
+              val versionsBody = Json.parse(resp.body).validate[Seq[Version]]
+              println(versionsBody)
```
Remove println pls

AndreiMaksimenko replied:
Oh! Fixed.

Dmitrii-Shevchenko commented on services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala:
```diff
@@ -17,6 +17,7 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with ScalaFutures with M
   val gitLabProjectVersioning: GitLabProjectVersioning = new GitLabProjectVersioning(mockHttpClient, gitLabConfig)
   import ProjectContext._
   "GitLabProjectVersioning" when {
+
```
Remove free line

Dmitrii-Shevchenko commented on services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala:
```diff
@@ -38,19 +39,50 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with ScalaFutures with M
+      activeProject.copy(repository = gitLabConfig.idPath + activeProject.projectId.value)
+    lazy val dummyVersion: Version = TestProjectUtils.getDummyVersion()
   }
+
```
Same

GrigoriiBerezin commented on services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala:
```diff
@@ -38,19 +39,50 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with ScalaFutures with M
     }
+
+    "getProjectVersions" should {
+      def dummyVersionsJson: String = s"[${Json.stringify(Json.toJson(dummyVersion))}]"
```
Why def?

IaroslavTavchenkov replied:
> Why def?

why don't you like defs ?)

AndreiMaksimenko replied:
Cause we can crush on the first test, and we will not reach this value, so I think it's better to call a method and do not have a val in memory.

GrigoriiBerezin replied:
Than lazy val is more fitting for this

AndreiMaksimenko replied:
Deal! Fixed.

GrigoriiBerezin reviewed (changes requested):
Nice done. But I have some questions

AndreiMaksimenko commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -30,13 +30,34 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
+            else {
+              println(resp.body)
+              val versionsBody = Json.parse(resp.body).validate[Seq[Version]]
+              println(versionsBody)
```
If you mean a pattern matching, so I think it will be less readable.

Dmitrii-Shevchenko reviewed (approved):
Great work!

OstapStrashevskii reviewed (approved):
Nice work!

myazinn commented on repositories/src/main/scala/cromwell/pipeline/datastorage/dto/Project.scala:
```diff
@@ -52,7 +52,16 @@ object ProjectUpdateRequest {
+final case class Version(name: String, message: String, target: String, commit: Commit)
+object Version {
+  implicit val versionPlayFormat: OFormat[Version] = Json.format[Version]
+  implicit val versionsPlayListFormat: Reads[Seq[Version]] = Reads.seq(versionPlayFormat)
```
No need for a reader of an object sequence.
If you do have a formatter for a single object, `play-json` must be able to handle other things automatically.

AndreiMaksimenko replied:
Very strange but everything works now. Fixed.

myazinn commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -36,7 +36,26 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
-  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] = ???
+
+  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Seq[Version]] =
+    if (project.repository == null)
+      Future.failed(VersioningException("There is no repository in this project"))
```
As we discussed, you should not check it here. If you do have a `Project`, you may assume that all of his fields are filled correctly. Otherwise, it must be fixed by whoever created an invalid object.

AndreiMaksimenko replied:
Fixed. Added creation of project on adding to DB.

myazinn reviewed (changes requested):
Good work!

myazinn commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -36,7 +36,26 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
+            if (resp.status != StatusCodes.OK.intValue)
+              Left(VersioningException(s"Could not take versions. Response status: ${resp.status}"))
+            else {
+              val versionsBody = Json.parse(resp.body).validate[Seq[Version]]
+              Right(versionsBody.get)
```
Almost forgot.
You must check whether `versionsBody` contains `JsSuccess` or `JsError` using pattern matching instead of simple `.get`. If it contains `JsError` then you must return `Left` instead of `Right`.

AndreiMaksimenko replied:
Fixed. But I ask you to check, I use "try" and "as" structures. Is it ok?

myazinn replied:
Good enough, but in my opinion, you can use `.validate` + `.fold` or `.validate` + pattern matching to handle this situation without throwing an exception. Take a look at the implementation of the `.as` method. You can do a similar thing.

AndreiMaksimenko replied:
Fixed to validate.

myazinn commented on services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala:
```diff
@@ -38,13 +38,41 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with ScalaFutures with M
     }
+
+    "getProjectVersions" should {
+      lazy val dummyVersionsJson: String = s"[${Json.stringify(Json.toJson(dummyVersion))}]"
```
You may leave it as is, but I think the following code is a bit better
`Json.stringify(Json.toJson(Seq(dummyVersion))`
Also, no need for a `lazy` modifier.

AndreiMaksimenko replied:
Lazy removed.

AndreiMaksimenko committed:
feat: adding repo creation on new project

Add Versioning to Project service. Refactor project creation.
On adding new project repositry will be created automaticly.
Add Repositry class to Project. Add withRepository method to Project
Refactor all the places where the repository was used

myazinn reviewed (approved):
Nice work!

ilyaonishenko commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -20,7 +18,8 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
 
   override def createRepository(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] = {
-    def projectWithRepository(repositoryFullPath: String): Project = project.copy(repository = repositoryFullPath)
+    def projectWithRepository(repositoryFullPath: String): Project =
+      project.copy(repository = Some(Repository(repositoryFullPath)))
```
It should be part of `Project` entity in my opinion. 
So it should be defined inside `Project` case class with the name `withRepository`

AndreiMaksimenko replied:
Done. Please check, did I understand you right.

ilyaonishenko replied:
@AndreiMaksimenko It should take Option[String]. 
In your case if we pass there, for example, `null` it won't be processed nicely. (Because of the `Some` )

myazinn replied:
@ilyaonishenko are you sure that it makes sense here? If `repositoryFullPath` is `null`, then we will have the following
`Option(Repository(null))`
not
`Option(null)`
Therefore, `Option` will not save us anyway.

AndreiMaksimenko replied:
Hmmm... it is strange I think. In that case of null who guarantees that there will not be Some[Null] in that Option - and we will have the same situation.

AndreiMaksimenko replied:
I can add one more method withNoRepositry() - to create empty repo Project.

AndreiMaksimenko replied:
Fixed

ilyaonishenko commented on services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala:
```diff
@@ -33,25 +32,48 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with ScalaFutures with M
       "return new active Project with 201 response" taggedAs Service in {
         val project = projectWithRepo
-        when(request(activeProject)).thenReturn(Future.successful(Response(StatusCodes.Created.intValue, "", Map())))
+        when(request(activeProject)).thenReturn(Future.successful(Response(HttpStatusCodes.Created, "", Map())))
```
It's better to create a value for it, for example `emptyBody`  and define it as "". Even better is to add some sense to this definition and give up idea of empty strings usage.

AndreiMaksimenko replied:
Fixed! I made EmptyBody - cause it is constant.

iliaonishchenko reviewed (changes requested):
Looks almost ready to me.  
There're some cosmetic changes need to be done.

AndreiMaksimenko commented on services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:
```diff
@@ -36,7 +36,26 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
-  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] = ???
+
+  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Seq[Version]] =
+    if (project.repository == null)
```
It was already fixed!
